### PR TITLE
Make sure subdomain setup is properly called + Add Framework level support constant subdomain materials

### LINKED ISSUE #9455: New interface functions in FEProblem are needed
### Description of the enhancement or error report
These new functions are used to check if a material property is defined on a block or a boundary.

### Rationale for the enhancement or information for reproducing the error
This allows integration of material properties over the entire domain or boundary without the properties being defined everywhere.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New interface function.

### LINKED ISSUE #9474: Current element pointer is invalid in subdomainSetup
### Description of the enhancement or error report
The current element pointer has not been set when MOOSE calls `SetupInterface::subdomainSetup()`, which is available on all elemental objects, such as `Kernel`s, `AuxKernel`s, `Material`s, etc.

### Rationale for the enhancement or information for reproducing the error
This is potentially a bug. See the discussion in #9456.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
No application is currently using the current element pointer in `subdomainSetup`, so no impact. The sooner we fix this, the better.

### PR DESCRIPTION
**Pull request by YaqiWang**
To close #9474. Two minor interface functions are added in `SubProblem`, which also closes #9455.

### CODE REVIEW

**Comment by YaqiWang** [comment]
@permcody I am so frustrated on trying to understand the exodus output. I must be ignoring something and can you give me a quick answer?

The four elements are assigned with blocks:
```
3 4
1 2
```
And the material property `prop1` is defined on block `1 2 4`, but I see the variable `var1` with ParaView has
```
1 0
1 1
```
which should be
```
0 1
1 1
```
I checked the function I added in this PR, it is working just fine. It seems there is a mismatch in between the variable dof numbering and element numbering.

**Comment by YaqiWang** [comment]
I did ignored one thing. When `subdomainSetup` is called, `_current_elem` has not be set yet...

**Comment by YaqiWang** [comment]
The second commit fixes the uninitialized current element in subdomain setup. I cannot think of a better way to do it.

**Comment by YaqiWang** [comment]
This is ready for review.

**Comment by YaqiWang** [comment]
@permcody Comments?

**Comment by YaqiWang** [comment]
About the second commit: we could let `SetupInterface` to hold old and current subdomain ID, (maybe also boundary ID). But that will require changes in all the threaded-loop-compute functions.

**Comment by rwcarlsen** [comment]
Why can't you just use ``SubProblem::getMaterialPropertyBlocks(std::string& prop_nae)``?  It returns a set that you can just do ``.count(block_id) > 0`` with.  Then we don't need to increase our API surface, expose internals of Assembly, etc.  What exactly is it you need to do? - not "what do you think you need to accomplish X" but rather "what is X"?

**Comment by YaqiWang** [comment]
The initial purpose of adding a separate function is to have a more efficient implementation to reduce the overhead inside an element loop. But I guess because I can do the check in `subdomainSetup`, the argument is weakened. But having another function is not a bad idea anyway.

Assembly or current subdomain ID is actually a separate issue. Currently inside `subdomainSetup`, we do not know the current subdomain ID. The second commit addresses this. As I said, it may not be the best way but surely the simplest way.

**Comment by permcody** [comment]
> The initial purpose of adding a separate function is to have a more efficient implementation to reduce the overhead inside an element loop.

By why are we doing checks inside of computation anyway? Materials and blocks should all be an up-front kind of check. We are doing a lot of these already in the `ActionSystem` execution and `initialSetup()`. Can these checks not be performed before the simulation starts?

**Comment by YaqiWang** [comment]
That will require I add as many block-restricted UOs as the materials and I will need to do thread-join and gathering for all these UOs together. It makes the code extremely awkward and hard to maintain.

**Comment by permcody** [comment]
I'm really not sure I follow. So you are using UserObjects to provide Materials? Let's take a step back, I know you are furiously working to try to optimize everything and you are using the framework in very novel ways but there may be better ways that we should address. Are you trying to avoid material calculations?

We'd like to reduce/combine loops a bit in the framework too, and I think we are poised to do a little of that. I also do think it's the panacea you are looking for to get several orders of magnitude that you might be seeking.

**Comment by permcody** [comment]
@friedmud and @andrsd - I just got done talking to @YaqiWang and I believe that there is indeed a bug (or perhaps a feature) in the `ThreadedElementLoopBase` class. The current element is not correct if you try to access the element in `subdomainSetup`:
https://github.com/idaholab/moose/blob/devel/framework/include/base/ThreadedElementLoopBase.h#L161-L164

The onElement() callback is called after the subdomain callback is called which updates the current element in Assembly. One could argue that you might want to look at the current element to get the subdomain since it might not be available in the class you are working with (i.e. Kernel), but the current element hasn't been set yet! There may be other ways of getting the right subdomain into the object but sadly it would still be out of sync with the current element even if it was available.

@YaqiWang proposed a fix here buried in with the rest of this PR, but I wasn't really liking it because it looks like he's reaching into Assembly to set the current element. It really feels like current element should probably be correct in that method. Take a look at the code here and let's decide on the best method to fix this issue. BTW - the description illustrates his test case and how he came to the conclusion that this was wrong to begin with.

**Comment by friedmud** [comment]
`_current_subdomain_id` or `_current_block` should be provided in all objects.  I also agree that `_current_elem` should be up to date when we call `subdomainSetup()`

**Comment by rwcarlsen** [comment]
Let's fix ThreadedElementLoopBase and avoid the setter  in Assembly.

**Comment by YaqiWang** [comment]
@rwcarlsen to make sure, you want to keep working on this?

**Comment by permcody** [comment]
@YaqiWang - I've been looking at the various compute loops and I don't think there's a better way of fixing this bug (i.e. avoiding the new Assembly interface). Many of the compute loops don't call any of the `prepare` or `reinit` methods on Assembly so attempting to factor that out and making that call beforehand might just be worse. 

I think the code you've put in there is acceptable based on the current design. Can you break it out from the rest of this PR and submit separately so we can get it reviewed?

**Comment by YaqiWang** [comment]
@permcody It is actually properly broken. The second commit is for fixing the `_current_elem` in subdomain change, except I also updated the gold exodus file of the test there. I can separate that exodus file out into the third commit if you want.

**Comment by permcody** [comment]
Well I'm not so sure I like you test objects 😄  
I think we can make a better test case, perhaps one that more closely reflects the need for this check in the first place. In either case, I'd like to deal with it separately. The fix should be reviewed and go in separately.

**Comment by YaqiWang** [comment]
How about this: I create a new issue for this subdomainChange thing and use change the test for checking this? The interface function is actually minor and trivial. The test just happens to use it.

**Comment by permcody** [comment]
Sounds good - please do it.

**Comment by YaqiWang** [comment]
I can also add a subdomain-wise constant material property using this feature to make sure not only aux kernels but also materials work fine with `subdomainSetup()`. Refer to #6979.

**Comment by YaqiWang** [comment]
There are something not quite right in those threaded loops. Investigating...

**Comment by YaqiWang** [comment]
@permcody Ready for your review.

**Comment by YaqiWang** [comment]
@permcody I have a small patch ready for #9479. But I'd like this to be reviewed and merged first. Thanks.

File: test/src/materials/SubdomainConstantMaterial.C
Comment (permcody) [inline test/src/materials/SubdomainConstantMaterial.C]: @idaholab/moose-team - Looks like Yaqi has found a way to avoid calculating properties at each quadrature point. This is fairly advanced user code and we have to decide whether or not we should support this. I'm not completely against the idea but if we are going to do this, it probably should be done in MOOSE and supported as a first-class capability. It took me a minute to figure out why he was resizing the scalar property holder and looping to max qps... Anyway, this is what this code looks like.

File: framework/include/base/ThreadedElementLoopBase.h
Comment (permcody) [inline framework/include/base/ThreadedElementLoopBase.h]: As far as fixing the value of _current_elem when we change subdomains, this is fine.

File: test/src/materials/SubdomainConstantMaterial.C
Comment (permcody) [inline test/src/materials/SubdomainConstantMaterial.C]: @YaqiWang - you definitely don't want to use `std::map::at()` This method throws a standard exception which won't get caught by our code that looks for `MooseException()`, it may not abort a parallel run in the most graceful manner. Instead you should just assert (optimized out in opt) that the value is there and access directly.

Comment (YaqiWang) [inline test/src/materials/SubdomainConstantMaterial.C]: Actually, this is exactly what @friedmud mentioned in #6979.

Comment (permcody) [inline test/src/materials/SubdomainConstantMaterial.C]: Yes it is indeed. However, if this is what we want, and it sounds like Derek is fine with this, we just need to build it into MOOSE. I don't think we want the users to have to size their own storage structures and do their own `_qp` loops. In other words, this code (as it sits) should go into MOOSE in "SubdomainConstantMaterial" or something like that. So the question is, what should the interface look like? Should we just have a single callback that's called "computeSubdomainMaterial" that just has you return a single value? Are you planning on having different values per quadrature point or anything crazy like that? Let's get this design right!

Comment (permcody) [inline test/src/materials/SubdomainConstantMaterial.C]: This is why I wanted to get the bug fix in separately. This constant material thing should probably be in a separate PR.

Comment (YaqiWang) [inline test/src/materials/SubdomainConstantMaterial.C]: My previous comments is regarding to the resize but not the use of `at`. Sorry for the confusion.

Comment (YaqiWang) [inline test/src/materials/SubdomainConstantMaterial.C]: Yes, we can have `Material::computeSubdomainQpProperty()`, which is called by `Material::subdomainSetup()`. Like what we did with `Material::_constant_on_elem`, we then copy the property on qp=0 to other qps. Is this something we want to support from now on?

Comment (YaqiWang) [inline test/src/materials/SubdomainConstantMaterial.C]: Because the material will be constant over the entire subdomain, we do not need to consider having different values per quadrature point.

Comment (YaqiWang) [inline test/src/materials/SubdomainConstantMaterial.C]: Also here we have to use `getMaxQps` because potentially the number of qps on elements of the subdomain varies with elements.

**Comment by YaqiWang** [comment]
I added three more commits to demonstrate the idea.

**Comment by friedmud** [comment]
Don't use my name to justify hacking the current API! 😛

I actually don't think this needs a separate base-class. I think this
should just be an option for any Material. If that material is
`const_on_sundomains = true` then MOOSE will do something similar to what's
being done here automatically.

In this way you can choose at runtime whether or not properties should be
evaluated like this for each Material.

**Comment by dschwen** [comment]
...and the value you get is just the value at that random element, which happens to be first in loop hitting the new subdomain ?

**Comment by dschwen** [comment]
Would that even be parallel reproducible? I think a new base class is the better choice here.

**Comment by permcody** [comment]
I can see it both ways. On one hand just adding an option could make this work with existing materials, but if people use any spatial information to calculate the material property it would be logically wrong and that would be too easy for a user to do. On the other hand, if we go with a new base class, what kind of coupling would be allowed in these constant subdomain materials? Could a user still couple to other materials, nonlinear variables etc? Seems a little dangerous. @YaqiWang - What kinds of information are you using to calculate your properties? I'm still trying to understand the requirements.

**Comment by YaqiWang** [comment]
@dschwen We set the current elem only for getting the current subdomain ID. Variables, shape functions have not been evaluated at all and they can not be used in subdomain setup.

**Comment by YaqiWang** [comment]
I feel that I am dragged into a hole. But fortunately the hole is not too deep. The story of this PR is:

1. First, I added two minor interface functions in `Subproblem` (see the first commit) for #9455, and I then added a test to make sure these functions are indeed working properly.
2. In this test, I used `AuxKernel::subdomainSetup()` and noticed the bug #9474, i.e. the current subdomain ID is not available, or the current element is lagged in the function.
3. So I went to add the issue #9474 and added the second commit to fix it suggested by @permcody. Besides testing `AuxKernel::subdomainSetup()`, I feel I should let the test also test `Material::subdomainSetup()`.
4. Then I realized the issue #9479. Every material has three copies in MOOSE for element interior, element face and neighboring element face. `subdomainSetup()` of neighboring materials needs to be called differently. Besides this, few threaded loops do not call subdomain setup at all. The third and fourth commits addressed these.
5. During fixing issue #9479, I found #9482 although I did not try to create a separate test to show the issue.
6. Finally I realized that once these `subdomainSetup()` fixes are in place, #6979 can be resolved as a by-product, which was in the last two commits.

**Comment by YaqiWang** [comment]
I am willing to work on this because I can see a potential usage in MAMMOTH. Sometimes, depletion can be done on subdomain-wise instead of on every quadrature point to same the computing resources. In such a case, variables (fluxes) need to be integrated over the subdomain first and then to be used by the material in subdomainSetup. So basically everything used in subdomainSetup must be subdomain-wise.

**Comment by dschwen** [comment]
@YaqiWang what about doing the depletion in an element user object. It would allow you to integrate and perform a per subdomain calculation. Then a simple material would just pull in the value from the UO. There would be very little overhead.

**Comment by YaqiWang** [comment]
@dschwen it is doable, but you will then have to have: pp(to integrating fluxes) <= UO (depletion) <= material, a very complicated dependency. If we can have something like `subdomainEnd()`, we can merge the pp and UO together and make things simpler. Even constant material is not needed we need to make `subdomainSetup()` properly called.

**Comment by YaqiWang** [comment]
Oh I see what you mean, depletion calculations can be done in `finalize`. Yes, that is actually a better idea, because we can control when depletion calculations is done easily.

**Comment by YaqiWang** [comment]
Comments?

**Comment by permcody** [comment]
Oh sorry, I didn't realize you had posted more code. I'll get to it when I'm done with my current ask.

**Comment by permcody** [comment]
@YaqiWang - I'll review this today if it's ready. Does the failed build require a patch to your apps?

**Comment by YaqiWang** [comment]
Thanks @permcody . Yes, it requires a trivial input parameter change.

File: framework/src/materials/Material.C
Comment (permcody) [inline framework/src/materials/Material.C]: @andrsd, @friedmud - Just getting back around to this whole constant subdomain material property issue. As you can see, @YaqiWang has taken a stab and this and it's much better than before. This is a pretty fundamental change though to the framework so I want to make sure you two buy off before we go any further.

File: framework/src/materials/Material.C
Comment (permcody) [inline framework/src/materials/Material.C]: I would prefer to see a class level enum for these cases. There's a templated getter on MooseEnum so you can pull out the enum type from MooseEnum to the native enum. It's a little cleaner than these magic numbers.

**Review by permcody — COMMENTED** [review]
I think this looks pretty good given the current state of MOOSE. I still would have liked that you fixed the current element problem separately from the constant subdomain properties fix but I suppose it's fine going in altogether. Just waiting for buy-off from the other devs on this.

**Comment by andrsd** [comment]
Currently the fix to have `_current_elem` set during `subdomainSetup()` is done by adding an extra setter on `Assembly`.

The current work flow in `ThreadedLoop...` is:
```
      _old_subdomain = _subdomain;
      _subdomain = cur_subdomain;

      if (_subdomain != _old_subdomain)
        subdomainChanged();

      onElement(elem);
```
For example, in `ComputeResidualThread::onElement` we have:
```
  _fe_problem.prepare(elem, _tid);
  _fe_problem.reinitElem(elem, _tid);
```
so, I think we should move `_fe_problem.prepare(elem, _tid);` into `ComputeResidualThread::subdomainChanged` before we call `_fe_problem.subdomainSetup(_subdomain, _tid);`. Obviously we would patch the other `ThreadedLoop`s in a similar way.

This patch will introduce 2 ways how current element is set and I'd like to avoid that.

**Comment by permcody** [comment]
@andrsd - I thought about moving the call to avoid the 2 ways problem, but it's a little awkward to move the `prepare()` method to `subdomainChanged()` because we don't have an element range to work on there to get the element from in the first place. Yeah, I really dislike this patch because it feels like a hack, but I couldn't see a better "clean" way. 

Maybe call subdomain setup at the beginning of the range operation? Still not clean.

**Comment by YaqiWang** [comment]
I think the logic in `ThreadedElementLoopBase` is fine and makes sense. Now I tend to the another way I mentioned: add two member variables, `_current_subdomain`, `_old_subdomain` in `SetupInterface` , when `subdomainChanged` is called in  `ThreadedElementLoopBase`, MOOSE will loop through all objects derived from `SetupInterface` and set the two member variables. To do this, we need to let MOOSE maintain a set of `SetupInterface` pointers.

**Comment by andrsd** [comment]
@permcody Looking at that again today, the major problem with my proposition is the prepare would be called only when subdomain changed. That's obviously wrong. So, now I am thinking about adding `preElement()`, similarly like we have `postElement()`.

**Comment by YaqiWang** [comment]
The code put in `preElement` might not being needed for all subdomains, although I am not against adding  `preElement`. I think it should not be used for serving this. `SetupInterface` calls all `subdomainSetup`, thus I think the most correct way is to change `SetupInterface`. This cost for setting subdomain ID should be negligible and if it it in place, we can avoid calls like `elem->subdomain_id()`.

File: framework/include/base/ThreadedElementLoopBase.h
Comment (YaqiWang) [inline framework/include/base/ThreadedElementLoopBase.h]: We could potentially enter here too many times if we have lots of blocks.

File: framework/src/materials/Material.C
Comment (YaqiWang) [inline framework/src/materials/Material.C]: It is actually not that magic, because when the MooseEnum is created, I explicitly specify `none=0 element=1 subdomain=2`. This can reduce the amount of code. But if you prefer native enum, I am willing to make changes.

Comment (permcody) [inline framework/src/materials/Material.C]: I mean "magic" as in an undocumented constant. Yes I can see that it's in the MOOSE Enum. Lately, we've gotten into the habit of creating a class scoped enum that we dump the MooseEnum into so that it's clean. Take a look at how this works:
https://github.com/idaholab/moose/blob/devel/framework/src/postprocessors/NumDOFs.C#L35

Comment (YaqiWang) [inline framework/src/materials/Material.C]: oh, i see. It is clean. I will use it later after we settle down the design.

**Comment by friedmud** [comment]
Just a heads up: I would like to hold off on merging this so we can have a
good discussion about all of these potential API changes at the tiger
team...

**Comment by YaqiWang** [comment]
@friedmud We had a discussion here and I will put in what we discussed.

**Comment by permcody** [comment]
@friedmud, @andrsd, @YaqiWang and I all talked about this quite a bit. We decided that we'd add a new "preElement" method in `ThreadedElementLoopBase` that we could use to set the current subdomain ID (a new field in `Assembly`. Actually, `Assembly` would be responsible for setting its own subdomain ID so it remains encapsulated). We discussed a few other options, such as moving `prepare` around but decided that this was probably the most semantically correct and least intrusive.

This is much cleaner than @YaqiWang's original proposal of adding new setters which could muck with the internals of Assembly.

File: framework/src/materials/Material.C
Comment (YaqiWang) [inline framework/src/materials/Material.C]: Actually, because I want to keep the name `none`, `element`, `subdomain`, which are not good for native enum (not capitalized, too common), I think the current way is much simpler...

**Comment by YaqiWang** [comment]
@permcody @andrsd Here are the commits based on our discussion.

File: framework/src/materials/Material.C
Comment (permcody) [inline framework/src/materials/Material.C]: Precisely why you need to use C++11 "class" enums. Then you'd refer to the values as "enum_type::value_name" (e.g. `ConstantType::subdomain`).

File: framework/include/base/Assembly.h
Comment (permcody) [inline framework/include/base/Assembly.h]: Probably should be marked `const` so it can be called from a const object later.

File: framework/include/base/Assembly.h
Comment (permcody) [inline framework/include/base/Assembly.h]: Mark `const`

File: framework/include/materials/Material.h
Comment (permcody) [inline framework/include/materials/Material.h]: Oh yeah, you are using a raw `int` here. Really, you should change this to an enum type.

File: framework/src/base/Assembly.C
Comment (permcody) [inline framework/src/base/Assembly.C]: 👍

File: framework/src/base/Assembly.C
Comment (permcody) [inline framework/src/base/Assembly.C]: 👍

File: test/src/materials/SubdomainConstantMaterial.C
Comment (permcody) [inline test/src/materials/SubdomainConstantMaterial.C]: This is the hard way, it's better to just use commas between each argument to `mooseError`, then you don't have to use conversion functions or anything else. We don't like `std::to_string()` in general because it doesn't format floats to double precision value, so we are trying to keep it out of the framework.

**Review by permcody — CHANGES_REQUESTED** [review]
This is looking pretty good. Just a few minor comments.

File: framework/include/base/Assembly.h
Comment (YaqiWang) [inline framework/include/base/Assembly.h]: right.

File: framework/src/materials/Material.C
Comment (YaqiWang) [inline framework/src/materials/Material.C]: If I have a variable in `Material` with name `subdomain` and I forget `ConstantType`, what will happen?

**Comment by YaqiWang** [comment]
Fixed the failure in debug mode due to unnecessary setting neighbor to null in few functions in Assembly. Addressed @permcody 's comments. App tests will fail due to a trivial input parameter name change.

**Comment by YaqiWang** [comment]
We need to create a benchmark with lots of blocks and check if this PR degrades the performance or not. @rwcarlsen might be interested in if he is working on benchmarking.

**Comment by YaqiWang** [comment]
@permcody those asserts in Assembly are causing the failure in debug mode. The last commit is used to fix the failure for framework. Apparently more are needed in module. Now I tend to replace those assert with subdomain ID assignment. Thoughts?

**Comment by permcody** [comment]
Looks like `reinitAtPhysical` needs a little fixup.

**Comment by YaqiWang** [comment]
Yes, looks like all from `NonlinearSystemBase::constraintResiduals()`. That subroutine needs to be converted to a threaded loop. `subdomainSetup` is not used at all. I am going to add a `subdomainSetup` in `Constraint` with final keyword and issuing a moose error saying this should be never called.

**Comment by YaqiWang** [comment]
Hopefully the last commit fixes all debug runs. I had few failing tests on my mac, but they could be caused by the different version of PETSc I am using. Let us see what Civet tells us.

**Comment by YaqiWang** [comment]
@permcody This is good now. I decided to keep that assert in `Assembly`. @friedmud please review it as you wish. It does not have too many changes. We decided to let `Assembly` hold the current subdomain ID and to add `preElement` and `preInternalSide` to set the current subdomain ID.

**Review by permcody — APPROVED** [review]
This looks fine to me. @friedmud - This monkeys around with MOOSE internals quite a bit. Make sure you don't have any issues with this.

File: framework/src/materials/Material.C
Comment (permcody) [inline framework/src/materials/Material.C]: 👍

**Review by permcody — DISMISSED** [review]

**Comment by YaqiWang** [comment]
@permcody Want me to rebase this? or it can be merged?

**Comment by permcody** [comment]
It was ready. Work with one of the guys on site to get it in. I'm on travel.

**Comment by YaqiWang** [comment]
@jwpeterson do you care to merge this in?

File: framework/include/materials/Material.h
Comment (andrsd) [inline framework/include/materials/Material.h]: All our enums are UPPER CASE.

File: framework/src/base/FEProblemBase.C
Comment (andrsd) [inline framework/src/base/FEProblemBase.C]: I know this is a copy'n'paste, but this comment is useless.

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by YaqiWang** [comment]
@andrsd addressed your comments and rebased. Once this is merged, I will send in a MR in rattlesnake. Thanks.

**Review by andrsd — APPROVED** [review]

**Comment by brianmoose** [comment]
@YaqiWang don't forget to do mammoth and sabertooth as well...

**Comment by YaqiWang** [comment]
Ahoh, will do.

**Pull request merged**
